example/test_problem/Hydro/CCSN/python_script/convert_progenitor_form.py

```diff
@@ -0,0 +1,151 @@
+        i_start = i_end
+        i_end   = i_start + other_column_size
+        header_append = header[i_start:i_end].strip()
+        headers.append( header_append )
+elif ( form == MESA ):
+    for i in range(5):
+       f.readline()
+    headers = f.readline().split()
+
+
+# find and read targeting variables (radius, density, temperature, radial velocity,
+#                                    electron fraction, pressure, and angular velocity )
+nvars = 7
+TargetCols = np.empty( nvars, dtype=np.int )
+if   ( form == WHS  ):
+    for i in range( len(headers) ):
+       if ( headers[i] == 'cell outer radius'     ): TargetCols[0] = i
+       if ( headers[i] == 'cell density'          ): TargetCols[1] = i
+       if ( headers[i] == 'cell temperature'      ): TargetCols[2] = i
+       if ( headers[i] == 'cell outer velocity'   ): TargetCols[3] = i
+       if ( headers[i] == 'cell Y_e'              ): TargetCols[4] = i
+       if ( headers[i] == 'cell pressure'         ): TargetCols[5] = i
+       if ( headers[i] == 'cell angular velocity' ): TargetCols[6] = i
+elif ( form == MESA ):
+   for i in range( len(headers) ):
+       if ( headers[i] == 'logR'                  ): TargetCols[0] = i
+       if ( headers[i] == 'density'               ): TargetCols[1] = i
+       if ( headers[i] == 'temperature'           ): TargetCols[2] = i
+       if ( headers[i] == 'velocity'              ): TargetCols[3] = i
+       if ( headers[i] == 'ye'                    ): TargetCols[4] = i
+       if ( headers[i] == 'pressure'              ): TargetCols[5] = i
+       if ( headers[i] == 'omega'                 ): TargetCols[6] = i
+
+
+# GAMER-formatted array
+GAMER_DATA_FORM = np.empty( (0, nvars), dtype=np.float )
+variables       = np.empty( nvars,      dtype=np.float )
+
+outer_radius = 0.0
+outer_vel    = 0.0
+for line in f:
+    if ( form == WHS  and  line[0:first_column_size].strip() == 'wind:' ): break
+    for idx, icol in enumerate(TargetCols):
+        variables[idx] = float( line.split()[icol] )
+#       convert cell outer variables to cell centered values (radius, velocity)
+        if ( form == WHS  ):
+            if ( idx == 0 ):
+                cell_centered_radius = ( outer_radius + variables[idx] ) / 2.0
+                outer_radius         = variables[idx]
+                variables[idx]       = cell_centered_radius
+            if ( idx == 3 ):
+                cell_centered_vel    = ( outer_vel + variables[idx] ) / 2.0 
+                outer_vel            = variables[idx]
+                variables[idx]       = cell_centered_vel
+#       unit conversion
+        if ( form == MESA ):
+            if ( idx == 0 ):
+                variables[idx] = 6.977e10 * np.power( 10, variables[idx] ) # unit conversion: radius
+            if ( idx == 3 ):
+                variables[idx] = 1e2 * variables[idx]                      # unit conversion: velocity
+
+    GAMER_DATA_FORM = np.vstack( [GAMER_DATA_FORM, variables] )
```

Comment:
- Why not do `line.split()` only once before for loop?
- For line 111-117, use `var[idx], outer = 0.5 * (outer + var[idx]), var[idx]`. This is not C/C++
- Define the value for unit conversion as global variables at beginning
- `np.vstack` is time-consuming. Why not use `list` to store the data, and convert to numpy array later. Also the conversion can be done once all the data are loaded
- If the input file has the same format for each type, why not use `np.genfromtxt` to load all the data at once, and manipulate it later?

Reply:
- Sorry, I don't get the point. could you explain more about this?
- I changed them.
- Now, I defined the values at the beginning.
- It might take more but the whole script is just finished in a second. I think it is okay to use `np.vstack`. I just personally prefer to use `numpy` than `list`.
- I am just not familiar with this. If you think this is better you can freely modify this part.

Reply:
I updated this to use `line.split()` only once before the for loops.